File: backend/IMIS/IMIS.Domain/AuditProgramme.cs
using Base.Primitives;
using System;
using System.Collections.Generic;

namespace IMIS.Domain
{
    public class AuditProgramme : Entity<int>
    {
        public int Year { get; set; }
        public required string For { get; set; }
        public required string From { get; set; }
        public required string Purpose { get; set; }

        public ICollection<AuditProgrammeObjective> Objectives { get; set; } = new List<AuditProgrammeObjective>();

        public required string ScopeAndFreqAudit { get; set; }
        public required string InternalAuditSched { get; set; }
        public required string AuditPlanObjective { get; set; }
        public required string ScopeOfAudit { get; set; }

        public required string AuditCriteria { get; set; }
        public required string AuditMethodology { get; set; }
        public required string SelectionAndEvaluationOfAuditors { get; set; }
        public required string Reporting { get; set; }
        public required string VerificationOfPreviousNonconformities { get; set; }
        public required string AuditLimitations { get; set; }

        
        public int AuditStatusId { get; set; } = AuditStatusSeedIds.Draft;
        public AuditPlanStatus? AuditStatus { get; set; }

        public ICollection<AuditProgrammeStatusHistory> StatusHistory { get; set; } = new List<AuditProgrammeStatusHistory>();

        public DateTime CreatedDate { get; set; } = DateTime.UtcNow;
        public DateTime? LastModifiedDate { get; set; }
        public ICollection<AuditPlanApproval> Approvals { get; set; } = new List<AuditPlanApproval>();
        public ICollection<AuditPlan> AuditPlans { get; set; } = new List<AuditPlan>();
    }
}

File: frontend/imis/lib/audit/audit_programme/models/audit_programme.dart
import 'package:json_annotation/json_annotation.dart';
import 'audit_programme_objective.dart';
import 'package:imis/audit/audit_plan/models/audit_plan.dart';

part 'audit_programme.g.dart';

@JsonSerializable(explicitToJson: true)
class AuditProgramme {
  @JsonKey(defaultValue: 0)
  int id;

  @JsonKey(defaultValue: false)
  bool isDeleted;

  @JsonKey(defaultValue: '')
  String rowVersion;

  @JsonKey(defaultValue: 0)
  int year;

  @JsonKey(name: 'for', defaultValue: '')
  String forUser;

  @JsonKey(name: 'from', defaultValue: '')
  String fromUser;

  @JsonKey(defaultValue: '')
  String purpose;

  @JsonKey(defaultValue: [])
  List<AuditProgrammeObjective> objectives;

  @JsonKey(defaultValue: '')
  String scopeAndFreqAudit;

  @JsonKey(defaultValue: '')
  String internalAuditSched;

  @JsonKey(defaultValue: '')
  String auditPlanObjective;

  @JsonKey(defaultValue: '')
  String scopeOfAudit;

  @JsonKey(defaultValue: '')
  String auditCriteria;

  @JsonKey(defaultValue: '')
  String auditMethodology;

  @JsonKey(defaultValue: '')
  String selectionAndEvaluationOfAuditors;

  @JsonKey(defaultValue: '')
  String reporting;

  @JsonKey(defaultValue: '')
  String verificationOfPreviousNonconformities;

  @JsonKey(defaultValue: '')
  String auditLimitations;

  @JsonKey(name: 'auditPlan', defaultValue: [])
  List<AuditPlan> auditPlans;

  AuditProgramme({
    this.id = 0,
    this.isDeleted = false,
    this.rowVersion = "",
    this.year = 0,
    this.forUser = "",
    this.fromUser = "",
    this.purpose = "",
    this.objectives = const [],
    this.scopeAndFreqAudit = "",
    this.internalAuditSched = "",
    this.auditPlanObjective = "",
    this.scopeOfAudit = "",
    this.auditCriteria = "",
    this.auditMethodology = "",
    this.selectionAndEvaluationOfAuditors = "",
    this.reporting = "",
    this.verificationOfPreviousNonconformities = "",
    this.auditLimitations = "",
    this.auditPlans = const [],
  });

  factory AuditProgramme.fromJson(Map<String, dynamic> json) =>
      _$AuditProgrammeFromJson(json);

  get forField => null;

  Map<String, dynamic> toJson() => _$AuditProgrammeToJson(this);
}